src/util/SweptCollection.js

```diff
@@ -0,0 +1,129 @@
+ * <info>This can allow the size to grow larger than max size when none of the entries pass the predicate.</info>
+ * @property {Function|null} [sweepFunction=null] A custom function passed directly to `sweep()` every `sweepInterval`.
+ * See {@link [Collection#sweep](https://discord.js.org/#/docs/collection/master/class/Collection?scrollTo=sweep)}
+ * for the definition of this function. This overrides the build in function generated using the other properties.
+ * @property {number} [sweepLifetime=14400] How long an entry should stay in the collection
+ * before it is considered sweepable
+ * @property {string} [sweepLifetimeProperty='createdTimestamp'] The property of the entry to check when
+ * determining the entries lifetime.
+ * @property {boolean} [sweepArchivedOnly=false] Whether to only sweep entries with a property `archived: true`.
+ * E.g. for threads.
+ * @property {number} [sweepInterval=3600] How frequently, in seconds, to sweep the collection.
+ */
+
+/**
+ * A Collection which holds a max amount of entries and sweeps periodically.
+ * The first key is deleted if the Collection has reached max size.
+ * @extends {Collection}
+ * @param {SweptCollectionOptions|Iterable} [options=null] Options for constructing the swept collection.
+ */
+class SweptCollection extends Collection {
+  constructor(options = null) {
+    if (typeof options !== 'object') throw new TypeError('INVALID_TYPE', 'options', 'object or iterable', true);
+    const {
+      maxSize = -1,
+      maxSizePredicate = null,
+      sweepFunction = null,
+      sweepLifetime = 14400,
+      sweepLifetimeProperty = 'createdTimestamp',
+      sweepInterval = 3600,
+      sweepArchivedOnly = false,
+      client = null,
+      iterable = null,
+    } = options;
+
+    if (typeof maxSize !== 'number') throw new TypeError('INVALID_TYPE', 'maxSize', 'number');
+    if (maxSizePredicate !== null && typeof maxSizePredicate !== 'function') {
+      throw new TypeError('INVALID_TYPE', 'maxSizePreidcate', 'function');
+    }
+    if (sweepFunction !== null && typeof sweepFunction !== 'function') {
+      throw new TypeError('INVALID_TYPE', 'sweepFunction', 'function');
+    }
+    if (typeof sweepFunction !== 'function' && typeof sweepLifetime !== 'number') {
+      throw new TypeError('INVALID_TYPE', 'sweepLifetime', 'number');
+    }
+    if (typeof sweepFunction !== 'function' && typeof sweepLifetimeProperty !== 'string') {
+      throw new TypeError('INVALID_TYPE', 'sweepLifetimeProperty', 'number');
+    }
+    if (typeof sweepInterval !== 'number') throw new TypeError('INVALID_TYPE', 'sweepInterval', 'number');
+    if (typeof sweepFunction !== 'function' && typeof sweepArchivedOnly !== 'boolean') {
+      throw new TypeError('INVALID_TYPE', 'sweepArchivedOnly', 'boolean');
+    }
+    if (typeof client !== 'object') throw new TypeError('INVALID_TYPE', 'client', 'Client');
```

Comment:
Wouldn't be good here to use `!(client instanceof Client)` instead?

Reply:
Technically speaking the jsdoc and typings say that you pass a djs client, but that's not necessarily a requirement,  its just an easy way to describe what it wants. For typings I can make it better, I have no idea what to do for the docs though.

The idea is you could pass your own client with a setInterval method since this doesn't *have* to be used within djs, allowing it to be destroyed with your client (that's the whole reason for it anyways, when done internally, we want to destroy the intervals when the client is destroyed)

Reply:
client is now removed from this :)